File: src/memory/git.ts
// Git changelog for the memory repo (DESIGN §5: "Every backend write is a git commit → a literal
// changelog of what the manager has learned"). The source of truth is markdown on disk; git gives
// us history/rollback for free. We shell out via execFileSync (no git library dependency).

import { execFileSync } from "node:child_process";
import { existsSync, mkdirSync } from "node:fs";
import { join } from "node:path";

function git(dir: string, args: string[]): string {
  return execFileSync("git", args, {
    cwd: dir,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
}

/** Ensure `dir` exists and is a git repo with a committer identity (idempotent). */
export function ensureRepo(dir: string): void {
  mkdirSync(dir, { recursive: true });
  if (existsSync(join(dir, ".git"))) return;
  git(dir, ["init", "-q"]);
  // Local identity so commits never depend on the host's global git config.
  git(dir, ["config", "user.name", "little-living-apps manager"]);
  git(dir, ["config", "user.email", "manager@lila.local"]);
}

/**
 * Stage everything and commit if (and only if) there is a change. Returns true when a commit was
 * made. `git diff --cached --quiet` exits non-zero when there ARE staged changes, so the throw is
 * the signal to commit.
 */
export function commitAll(dir: string, message: string): boolean {
  git(dir, ["add", "-A"]);
  try {
    git(dir, ["diff", "--cached", "--quiet"]);
    return false; // exit 0 → nothing staged → nothing to commit
  } catch {
    git(dir, ["commit", "-q", "-m", message]);
    return true;
  }
}

/** Number of commits on HEAD (0 if the repo has no commits yet). Used by tests/diagnostics. */
export function commitCount(dir: string): number {
  try {
    return Number(git(dir, ["rev-list", "--count", "HEAD"]).trim()) || 0;
  } catch {
    return 0;
  }
}

File: src/memory/git.rs
//! Git changelog for the memory repo. Port of `src/memory/git.ts`: every backend write is a commit,
//! so history is a literal changelog of what the manager has learned. We shell out to `git` (no
//! native git library) — keeps the dependency tree light and the static binary musl-friendly.

use std::path::Path;
use std::process::Command;

use super::MemoryError;

/// Run a git subcommand in `dir`, returning stdout on success.
fn git(dir: &Path, args: &[&str]) -> Result<String, MemoryError> {
    let output = Command::new("git")
        .args(args)
        .current_dir(dir)
        .output()
        .map_err(|e| MemoryError(format!("git {args:?} failed to spawn: {e}")))?;
    if output.status.success() {
        Ok(String::from_utf8_lossy(&output.stdout).into_owned())
    } else {
        Err(MemoryError(format!(
            "git {args:?} exited {}: {}",
            output.status,
            String::from_utf8_lossy(&output.stderr).trim()
        )))
    }
}

/// Ensure `dir` exists and is a git repo with a committer identity (idempotent).
pub fn ensure_repo(dir: &Path) -> Result<(), MemoryError> {
    std::fs::create_dir_all(dir).map_err(|e| MemoryError(format!("mkdir {dir:?}: {e}")))?;
    if dir.join(".git").exists() {
        return Ok(());
    }
    git(dir, &["init", "-q"])?;
    // Local identity so commits never depend on the host's global git config.
    git(dir, &["config", "user.name", "little-living-apps manager"])?;
    git(dir, &["config", "user.email", "manager@lila.local"])?;
    Ok(())
}

/// Stage everything and commit iff there is a change. Returns `true` when a commit was made.
/// `git diff --cached --quiet` exits non-zero when there ARE staged changes — that's the commit
/// signal.
pub fn commit_all(dir: &Path, message: &str) -> Result<bool, MemoryError> {
    git(dir, &["add", "-A"])?;
    let clean = Command::new("git")
        .args(["diff", "--cached", "--quiet"])
        .current_dir(dir)
        .status()
        .map_err(|e| MemoryError(format!("git diff failed: {e}")))?
        .success();
    if clean {
        return Ok(false); // nothing staged
    }
    git(dir, &["commit", "-q", "-m", message])?;
    Ok(true)
}

/// Number of commits on HEAD (0 if no commits yet). Used by tests/diagnostics.
pub fn commit_count(dir: &Path) -> i64 {
    git(dir, &["rev-list", "--count", "HEAD"])
        .ok()
        .and_then(|s| s.trim().parse::<i64>().ok())
        .unwrap_or(0)
}
